dhall/src/Dhall/Pretty/Internal.hs

```diff
@@ -171,107 +171,76 @@ prettySrcExpr = prettyCharacterSet Unicode
-        x = 1 in x
-    >>> let emptySrc = Src unusedSourcePos unusedSourcePos "      "
-    >>> "let" <> " " <> renderSrc id (Just emptySrc) <> "x = 1 in x"
-    let x = 1 in x
-    >>> "let" <> " " <> renderSrc id Nothing <> "x = 1 in x"
-    let x = 1 in x
--}
-renderSrc
-    :: (Text -> Text)
-    -- ^ Used to preprocess the comment string (e.g. to strip whitespace)
-    -> Maybe Src
-    -- ^ Source span to render (if present)
-    -> Doc Ann
-renderSrc strip (Just (Src {..}))
-    | not (Text.all isWhitespace srcText) =
-        renderComment (strip srcText)
-renderSrc _ _ =
-    mempty
-
-{-| Render a comment.
-
-    Any preprocessing, such as whitespace stripping, needs to be handled by the
-    caller, see e.g. 'renderSrc'.
-
-    See the documentation for 'renderSrc' for examples.
--}
-renderComment :: Text -> Doc Ann
-renderComment text =
-    Pretty.align (Pretty.concatWith f newLines <> suffix)
+normalizeMultiComment :: MultiComment -> (Bool, [Text])
+normalizeMultiComment (MultiComment comments) =
+    ( any isBlockComment comments
+    , atLeastOneLine . stripEmptyLines $ concatMap toLines comments
+    )
   where
-    horizontalSpace c = c == ' ' || c == '\t'
-
-    suffix =
-        if Text.null text || Text.last text == '\n'
-        then mempty
-        else " "
-
-    oldLines = Text.splitOn "\n" text
-
-    spacePrefix = Text.takeWhile horizontalSpace
-
-    commonPrefix a b = case Text.commonPrefixes a b of
-        Nothing        -> ""
-        Just (c, _, _) -> c
-
-    sharedSpacePrefix []       = ""
-    sharedSpacePrefix (l : ls) = foldl' commonPrefix (spacePrefix l) ls
-
-    blank = Text.all horizontalSpace
-
-    newLines =
-        case oldLines of
-            [] ->
-               []
-            l0 : ls ->
-                let sharedPrefix =
-                        sharedSpacePrefix (filter (not . blank) ls)
-
-                    perLine l =
-                        case Text.stripPrefix sharedPrefix l of
-                            Nothing -> Pretty.pretty l
-                            Just l' -> Pretty.pretty l'
-
-                in  Pretty.pretty l0 : map perLine ls
+    isBlockComment = \case
+        BlockComment _ -> True
+        LineComment _ -> False
+
+    stripStartEmptyLines = dropWhile Text.null
+
+    -- Remove leading and terminating empty lines
+    stripEmptyLines =
+          reverse
+        . stripStartEmptyLines
+        . reverse
+        . stripStartEmptyLines
+
+    -- When comment is empty, keep exactly 1 empty line
+    atLeastOneLine = \case
+        [] -> [ "" ]
+        xs -> xs
+
+    toLines = \case
+        BlockComment txt
+          | Just rest <- Text.stripPrefix "{-" txt
+          , Just body <- Text.stripSuffix "-}" rest
+          -> Text.strip <$> Text.lines body
+        LineComment txt
+          | Just body <- Text.stripPrefix "--" txt
+          -> [ Text.strip body ]
+        _ -> error "Dhall.Pretty.Internal.normalizeMultiComment: Unexpected comment"
```

Comment:
I suggest using `Dhall.Core.internalError` here and including the text of the comment in the error message